### ISSUE TITLE
Split rectangles are not considered for widescreen projection

### ISSUE THREAD

**Issue by klorfmorf** [rt64/rt64#28]
Split rectangles, by which I mean rectangles which have the same vertical span but are split horizontally, are not considered for any widescreen projection (`invRatioScale` not being set to `1.0f`).

This can occur in grids of texrects that make up a larger image where each rectangle is a **_square_** part of that image.

_Example:_
![Square Rectangle Grid](https://github.com/rt64/rt64/assets/169589138/ebba05d1-bf4d-407e-b5fc-8cb6d8cbeea6)

In the case of Mystical Ninja Starring Goemon (again, haha) this happens on the Konami logo, where the background is made up of square texrects, and on certain camera angles with the skybox (kind of hard to explain, but the skybox's texrects get split in two when the parts where the skybox texture starts repeating is visible on screen, essentially meaning the game is rendering two skyboxes split horizontally).

I wrote some quick-and-dirty code that fixes the issue, but I am sure someone more well versed in this codebase could implement it in a better way. If this has too big of a performance impact then maybe this could be implemented like the `fixLR` flag.

```c
case Projection::Type::Rectangle: {
    instanceDrawCall.type = InstanceDrawCall::Type::RegularRect;
    triangles.indexStart = call.meshDesc.rawVertexStart;
    fixedRect = fixRect(call.callDesc.rect, fbPair.scissorRect, p.fixRectLR);

    bool tileCopiesUsed = false;
    for (uint32_t t = 0; (t < call.callDesc.tileCount) && !tileCopiesUsed; t++) {
        tileCopiesUsed = drawData.callTiles[call.callDesc.tileIndex + t].tileCopyUsed;
    }

    // Check if there are any adjacent rectangles on the same span of height. 
    bool mergeRectangles = false;
    int32_t furthestLeft = 0;
    int32_t furthestRight = 0;

    // Only check this projection.
    for (uint32_t i = 0; (i < proj.gameCallCount); i++) {
        const GameCall &otherCall = proj.gameCalls[i];
        const FixedRect otherRect = fixRect(otherCall.callDesc.rect, fbPair.scissorRect, p.fixRectLR);

        if ((otherRect.uly == fixedRect.uly) && (otherRect.lry == fixedRect.lry)) {
            mergeRectangles = true;
            furthestLeft = fmin(furthestLeft, otherRect.ulx);
            furthestRight = fmax(furthestRight, otherRect.lrx);
        }
    }

    // There are adjacent rectangles on the same span of height which also span the whole width of the scissor.
    const bool coversScissorWidthMerged = mergeRectangles && (furthestLeft <= fbPair.scissorRect.ulx) && (furthestRight >= fbPair.scissorRect.lrx);

    // The call's scissor spans the whole width of the framebuffer pair scissor. The rect must not be using extended origins.
    const bool regularOrigins = (call.callDesc.rectLeftOrigin == G_EX_ORIGIN_NONE) && (call.callDesc.rectRightOrigin == G_EX_ORIGIN_NONE);
    const bool coversScissorWidth = regularOrigins && (fixedRect.ulx <= fbPair.scissorRect.ulx) && (fixedRect.lrx >= fbPair.scissorRect.lrx);
    if (tileCopiesUsed || coversScissorWidth || coversScissorWidthMerged) {
        invRatioScale = 1.0f;
    }
    else {
        horizontalMisalignment = p.horizontalMisalignment;
    }

    triangles.viewport = convertViewportRect(fixedRect, p.resolutionScale, p.fbWidth, invRatioScale, extOriginPercentage, horizontalMisalignment, call.callDesc.rectLeftOrigin, call.callDesc.rectRightOrigin);
    break;
}
```

**Before the patch:**
![Before Konami Logo](https://github.com/rt64/rt64/assets/169589138/feb15537-2dd2-467d-b68e-fcf0d0d3ae64)
![Before Skybox](https://github.com/rt64/rt64/assets/169589138/b4518dd2-3bc8-4d97-8d51-4ac0da0f576a)

**After the patch:**
![After Konami Logo](https://github.com/rt64/rt64/assets/169589138/4018ce14-f9a2-4b8c-86e3-17efaa314f03)
![After Skybox](https://github.com/rt64/rt64/assets/169589138/bf59b620-0f09-425f-a641-fc415dc140c2)

**Comment by DarioSamo** [comment]
This is an interesting point where I think I'd want to provide a better way to control this behavior through the extended GBI instead as in some games one behavior is preferable to the other. The proposed solution is certainly a bit heavy-handed but I can probably think of a better place to have it.

But I'll start by asking something from the second example: are you sure you want to stretch the sky like that? Wouldn't it make more sense in that case to patch the game to draw more of it to the sides? I realize this sounds impossible because you can't go out of bounds, but [the extended GBI will allow you to draw rects in negative coordinates to fix this](https://github.com/rt64/rt64/blob/main/include/rt64_extended_gbi.h).

I do think we can consider evaluating adding a simple command to the extended GBI to allow to basically control if rectangles should be auto-detected, stretched or maintain their 4:3 ratio though. Let me know what your ideal solution is in this case as it's a bit of a tricky area to come up with a generic solution at the renderer level, so we have to be a bit careful to not break any other solutions that rely on the auto-detection.

**Comment by klorfmorf** [comment]
I didn't know about the extended GBI, but now that I know I would definitely just patch the skybox to draw more using it. Though in situations where one hasn't really done enough reverse engineering those extended GBI commands for setting rectangles sound like a good idea.

**Comment by DarioSamo** [comment]
> I didn't know about the extended GBI, but now that I know I would definitely just patch the skybox to draw more using it. Though in situations where one hasn't really done enough reverse engineering those extended GBI commands for setting rectangles sound like a good idea.

Yeah, the basic idea behind these commands is to provide ways to either directly patch one command or to add a command that affects all other commands inside the region. A lot of that was used in Zelda for example to forcibly align parts of the HUD to a side of the screen without having to patch all the underlying functions.

I can certainly provide something that forces the algorithm to adapt or not for all rectangles in the spirit of the other commands if that is preferrable.

**Comment by klorfmorf** [comment]
> I can certainly provide something that forces the algorithm to adapt or not for all rectangles drawn afterwards in the spirit of the other commands if that is preferable.

Yeah that'd be useful, say I wanted to extend the skybox properly and not just stretch it, if some extended GBI command that forces the aspect ratio for a certain viewport (or something like that) existed then I could tell it to keep it at the original aspect ratio but by default it would just stretch it according to the heuristic (when RT64 is used as an emulator plugin and the user doesn't have any access to patches)

**Comment by DarioSamo** [comment]
We decided with Wiseguy to add a new GBI extended command to control the heuristic directly of any rects that draw in-between (Automatic, Original or Expand). I should be able to add this soon or feel free to take a stab at it yourself if you need it quickly.
